File: pset2/readability.c
/*
    Keith R Bergerstock
    krbergerstock@e4kountdown.com
    CS50 - readability

*/

#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>

#define MAX_BUF_SIZE 512
float calc_colman_laiu(int sentences, int words, int letters)
{
    float L = (float)letters / (float)words * 100.0;
    float S = (float)sentences / (float)words * 100.0;
    return 0.0588 * L - 0.296 * S - 15.8;
}

int count_characters(const char *buffer)
{
    int letters = 0;
    int words = 0;
    int sentences = 0;
    char u = 0;
    char *t = (char *)buffer;
    while (*t)
    {
        if (isalpha(*t))
        {
            letters++;
        }
        else if(*t == '!' || *t == '?1' || *t == '.')
        {
            sentences++;
            words++;
        }
        else if(*t == ',' || *t == '.')
        {
            words++;
        }
        else if(*t == ' ' && isalpha(u))
        {
            words++;
        }
        if (*t)
        {
            u = *t;
            t++;
        }
    }
    float cl = calc_colman_laiu(sentences, words, letters);
    int grade = (int)(cl + 0.5);
    return grade;
}


int main(void)
{
    char *buffer = calloc(MAX_BUF_SIZE,  sizeof(char));
    fputs("TEXT: ", stdout);
    fgets(buffer, MAX_BUF_SIZE, stdin);
    int grade = count_characters(buffer);
    if (grade < 1)
    {
        fputs("Before Grade 1\n", stdout);
    }
    else if (grade >= 16)
    {
        fputs("Grade 16+\n", stdout);
    }
    else
    {
        fprintf(stdout, "Grade %d\n", grade);
    }

    free(buffer);
    return 0;
}

File: pset6/readability/readability.py
# readability.py
#
# Keith R Bergerstock
# date: 2020.11.16

import sys

def inputString(prompt):
    """ returns a string from stdin """
    sys.stdout.write(prompt)
    sys.stdout.flush()
    tstr = sys.stdin.readline()
    return tstr.rstrip()

def calc_colman_laiu(sentences, words, letters):
    L = letters / words * 100.0
    S = sentences / words * 100.0
    return 0.0588 * L - 0.296 * S - 15.8


def count_characters(text):
    letters = 0
    words = 0
    sentences = 0
    t = ''
    for c in text:
        if c.isalpha():
            letters += 1
        elif c == '!' or c == '?' or c == '.':
            sentences += 1
            words  += 1
        elif c == ';' or c == ',':
            words += 1
        elif c == ' ' and t.isalpha():
            words += 1
        # store this pass to be used by the next pass            
        t = c    

    # print('sentences ',sentences)        
    # print(' words    ',words)
    # print('letters   ',letters)
    cl = calc_colman_laiu(sentences, words, letters)
    return int(cl + 0.5)


def main():
    txt = inputString("TEXT:")
    grade = count_characters(txt)
    if grade < 1:
        print('Before Grade 1')
    elif grade >= 16:    
        print('Grade 16+')
    else:
        print(f"Grade {grade} ")    


if __name__ == "__main__":
    main()
